# Spam Detector

In [19]:
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

## Retrieve the Data

The data is located at [https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv](https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv)

Dataset Source: [UCI Machine Learning Library](https://archive.ics.uci.edu/dataset/94/spambase)

Import the data using Pandas. Display the resulting DataFrame to confirm the import was successful.

In [20]:
# Import the data
data = pd.read_csv("https://static.bc-edx.com/ai/ail-v-1-0/m13/challenge/spam-data.csv")
data.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [21]:
data.columns

Index(['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
       'word_freq_our', 'word_freq_over', 'word_freq_remove',
       'word_freq_internet', 'word_freq_order', 'word_freq_mail',
       'word_freq_receive', 'word_freq_will', 'word_freq_people',
       'word_freq_report', 'word_freq_addresses', 'word_freq_free',
       'word_freq_business', 'word_freq_email', 'word_freq_you',
       'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
       'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
       'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
       'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
       'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
       'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
       'word_freq_original', 'word_freq_project', 'word_freq_re',
       'word_freq_edu', 'word_freq_table', 'word_freq_conference',


## Predict Model Performance

You will be creating and comparing two models on this data: a Logistic Regression, and a Random Forests Classifier. Before you create, fit, and score the models, make a prediction as to which model you think will perform better. You do not need to be correct! 

Write down your prediction in the designated cells in your Jupyter Notebook, and provide justification for your educated guess.

As an educated guess, I predict the RandomForest to perform better as that model leverages the wisdom of crowds by combining multiple decision trees trained on different subsets of data and features, thereby achieving better predictive performance and robustness compared to individual decision trees. This is very typical to that model, thence allowing its prediction to be better than the logistic regression which is very simplistic at its core.

## Split the Data into Training and Testing Sets

In [22]:
# Create the labels set `y` and features DataFrame `X`
X = data.copy()
X = X.drop(columns='spam')

# To create the target DataFrame, y
y = data['spam']

In [23]:
# Check the balance of the labels variable (`y`) by using the `value_counts` function.
y.value_counts()

spam
0    2788
1    1813
Name: count, dtype: int64

In [24]:
# Split the data into X_train, X_test, y_train, y_test
X_train, X_test, y_train, y_test = train_test_split(X, y)

In [25]:
X_train

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
1271            0.00               0.52           1.05           0.0   
1855            0.00               1.28           0.00           0.0   
2323            0.00               0.00           0.00           0.0   
3311            0.00               0.00           0.00           0.0   
1470            0.00               0.00           1.55           0.0   
...              ...                ...            ...           ...   
2781            0.00               0.00           0.00           0.0   
4287            0.00               0.00           0.00           0.0   
2714            0.03               0.03           0.00           0.0   
3527            0.09               0.09           0.28           0.0   
3205            0.00               0.00           0.60           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
1271           2.63            0.52              1.05                0.00   
1855           0.00            0.00              0.00                0.00   
2323           1.38            0.00              0.00                0.00   
3311           0.00            0.00              0.00                0.00   
1470           0.00            0.77              0.00                0.38   
...             ...             ...               ...                 ...   
2781           0.00            0.00              0.00                0.00   
4287           3.44            0.00              0.00                0.00   
2714           0.00            0.00              0.00                0.00   
3527           0.28            0.00              0.00                0.28   
3205           0.60            0.00              0.00                0.00   

      word_freq_order  word_freq_mail  ...  word_freq_conference  char_freq_;  \
1271             0.00            0.00  ...                   0.0        0.000   
1855             0.00            0.00  ...                   0.0        0.000   
2323             0.00            0.00  ...                   0.0        0.000   
3311             0.00            0.00  ...                   0.0        0.000   
1470             0.00            0.00  ...                   0.0        0.000   
...               ...             ...  ...                   ...          ...   
2781             0.00            0.00  ...                   0.0        0.000   
4287             0.00            0.00  ...                   0.0        0.000   
2714             0.06            0.09  ...                   0.0        0.024   
3527             0.00            0.00  ...                   0.0        0.014   
3205             0.00            0.00  ...                   0.0        0.000   

      char_freq_(  char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
1271        0.000        0.091        1.276        0.729        0.091   
1855        0.196        0.000        0.000        0.000        0.000   
2323        0.238        0.000        0.000        0.000        0.000   
3311        0.000        0.000        0.000        0.000        0.000   
1470        0.060        0.000        0.609        0.121        0.000   
...           ...          ...          ...          ...          ...   
2781        0.000        0.000        0.000        0.000        0.000   
4287        0.000        0.000        0.000        0.000        0.000   
2714        0.178        0.000        0.128        0.000        0.000   
3527        0.084        0.000        0.042        0.000        0.042   
3205        0.110        0.000        0.000        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
1271                       3.062                          19   
1855                       1.952                          10   
2323                       1.578                           4   
3311                       1.400                           3   
1470                       2.666              

In [26]:
X_train.columns

Index(['word_freq_make', 'word_freq_address', 'word_freq_all', 'word_freq_3d',
       'word_freq_our', 'word_freq_over', 'word_freq_remove',
       'word_freq_internet', 'word_freq_order', 'word_freq_mail',
       'word_freq_receive', 'word_freq_will', 'word_freq_people',
       'word_freq_report', 'word_freq_addresses', 'word_freq_free',
       'word_freq_business', 'word_freq_email', 'word_freq_you',
       'word_freq_credit', 'word_freq_your', 'word_freq_font', 'word_freq_000',
       'word_freq_money', 'word_freq_hp', 'word_freq_hpl', 'word_freq_george',
       'word_freq_650', 'word_freq_lab', 'word_freq_labs', 'word_freq_telnet',
       'word_freq_857', 'word_freq_data', 'word_freq_415', 'word_freq_85',
       'word_freq_technology', 'word_freq_1999', 'word_freq_parts',
       'word_freq_pm', 'word_freq_direct', 'word_freq_cs', 'word_freq_meeting',
       'word_freq_original', 'word_freq_project', 'word_freq_re',
       'word_freq_edu', 'word_freq_table', 'word_freq_conference',


In [27]:
y_train.dtypes

dtype('int64')

## Scale the Features

Use the `StandardScaler` to scale the features data. Remember that only `X_train` and `X_test` DataFrames should be scaled.

In [28]:
from sklearn.preprocessing import StandardScaler

# Create the StandardScaler instance
scaler = StandardScaler()

In [29]:
# Fit the Standard Scaler with the training data
X_train_scaled_n_fitted=scaler.fit_transform(X_train)
#X_test_scaled_n_fitted=scaler.fit_transform(X_test)

In [30]:
# Scale the training data
X_train_scaled_n_transformed = scaler.transform(X_train)
X_test_scaled_n_transformed = scaler.transform(X_test)

## Create and Fit a Logistic Regression Model

Create a Logistic Regression model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [31]:
# Train a Logistic Regression model and print the model score
from sklearn.linear_model import LogisticRegression
logistic_regression_model = LogisticRegression(random_state=1)
logistic_regression_model.fit(X_train_scaled_n_transformed, y_train)
print(f"Training Data Score: {logistic_regression_model.score(X_train_scaled_n_transformed, y_train)}")


Training Data Score: 0.9240579710144927


In [32]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = logistic_regression_model.predict(X_test_scaled_n_transformed)
# Review the predictions
testing_predictions

array([0, 0, 1, ..., 1, 0, 0], dtype=int64)

In [33]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, testing_predictions)
print("Accuracy Score:", accuracy)

Accuracy Score: 0.9331016507384883


## Create and Fit a Random Forest Classifier Model

Create a Random Forest Classifier model, fit it to the training data, make predictions with the testing data, and print the model's accuracy score. You may choose any starting settings you like. 

In [34]:
# Train a Random Forest Classifier model and print the model score
from sklearn.ensemble import RandomForestClassifier
random_forest_model = RandomForestClassifier(random_state=1)
random_forest_model.fit(X_train_scaled_n_transformed, y_train)

RandomForestClassifier(random_state=1)

In [35]:
# Make and save testing predictions with the saved logistic regression model using the test data
testing_predictions = random_forest_model.predict(X_test_scaled_n_transformed)

# Review the predictions
testing_predictions

array([0, 0, 1, ..., 1, 0, 0], dtype=int64)

In [36]:
# Calculate the accuracy score by evaluating `y_test` vs. `testing_predictions`.
accuracy = accuracy_score(y_test, testing_predictions)
print("Accuracy Score:", accuracy)


Accuracy Score: 0.9609035621198957


## Evaluate the Models

Which model performed better? How does that compare to your prediction? Write down your results and thoughts in the following markdown cell.


The random forest performed better because it yielded an accuracy score of .96 compared to the logistic regression model which yielded an accuracy score of .93. This corroborates with my earlier educated guess.